GunsOn/Program.cs

```diff
@@ -0,0 +1,117 @@
+﻿using Sandbox.ModAPI.Ingame;
+using SpaceEngineers.Game.ModAPI.Ingame;
+using System.Collections.Generic;
+
+namespace IngameScript {
+	partial class Program : MyGridProgram {
+		private const string GunsOn = "on";
+		private const string GunsOff = "off";
+		private const bool ReconfigureWhenTurnedOn = true;
+		private const bool ReconfigureWhenTurnedOff = false;
+
+		private readonly List<IMySmallGatlingGun> _smallGatlingGuns = new List<IMySmallGatlingGun>();
+		private readonly List<IMySmallMissileLauncher> _smallMissileLaunchers = new List<IMySmallMissileLauncher>();
+		private readonly List<IMyLargeConveyorTurretBase> _largeConveyoredGuns = new List<IMyLargeConveyorTurretBase>();
+		private readonly List<IMyLargeTurretBase> _largeNonConveyoredGuns = new List<IMyLargeTurretBase>();
+
+		private IEnumerator<bool> _stateMachine;
+
+		public Program() {
+			Runtime.UpdateFrequency = UpdateFrequency.None;
+		}
+
+		public void Main(string argument, UpdateType updateSource) {
+			if ((updateSource & UpdateType.Once) == UpdateType.Once) {
+				RunStateMachine();
+				return;
+			}
+			switch (argument) {
+				case "":
+				case null:
+					break;
+				case GunsOn:
+					_stateMachine = SetAllGunsTo(true, ReconfigureWhenTurnedOn).GetEnumerator();
+					break;
+				case GunsOff:
+					_stateMachine = SetAllGunsTo(false, ReconfigureWhenTurnedOff).GetEnumerator();
+					break;
+				default:
+					Echo("Invalid command.");
+					break;
+			}
+			RunStateMachine();
+		}
+
+		public void RunStateMachine() {
+			if (_stateMachine != null) {
+				if (_stateMachine.MoveNext() && _stateMachine.Current) {
+					Runtime.UpdateFrequency |= UpdateFrequency.Once;
+				} else {
+					_stateMachine.Dispose();
+					_stateMachine = null;
+				}
+			}
+		}
+
+		private IEnumerable<bool> SetAllGunsTo(bool targetState, bool configure) {
+			GridTerminalSystem.GetBlocksOfType(_largeConveyoredGuns);
+			if (configure) {
+				ConfigureLargeConveyoredGuns();
+			}
+			foreach (var gun in _largeConveyoredGuns) {
+				gun.Enabled = true;
+			}
+
+			yield return true;
+
+			GridTerminalSystem.GetBlocksOfType(_smallGatlingGuns);
+			foreach (var gun in _smallGatlingGuns) {
+				gun.Enabled = true;
+			}
+
+			yield return true;
+
+			GridTerminalSystem.GetBlocksOfType(_smallMissileLaunchers);
+			foreach (var gun in _smallMissileLaunchers) {
+				gun.Enabled = true;
+			}
+
+			yield return true;
+
+			GridTerminalSystem.GetBlocksOfType(_largeNonConveyoredGuns);
+			_largeConveyoredGuns.RemoveAll(g => g is IMyLargeConveyorTurretBase);
+			if (configure) {
+				ConfigureLargeUnconveyoredGuns();
+			}
+			foreach (var gun in _largeNonConveyoredGuns) {
+				gun.Enabled = targetState;
+			}
+
+			yield return false;
+		}
+
+		private void ConfigureLargeConveyoredGuns() {
+			foreach (var gun in _largeConveyoredGuns) {
+				gun.TargetCharacters = false;

```

Comment:
these are sensible defaults, and likely what I would use as well, but it might be nice (as a 2.0 feature) to have these configurable in custom data

Reply:
I am generally in favor of this, though I have performance concerns if the custom data needs to be parsed for every gun on every run.

That said, MDK-SE does have a helper to treat custom data fields as [slightly custom INI config files](https://github.com/malware-dev/MDK-SE/wiki/VRage.Game.ModAPI.Ingame.Utilities.MyIni).

... maybe using the programmable block's custom data instead of the individual guns (or was that your initial suggestion)?

Reply:
That was my suggestion, yes. I would assume you would hold it statically after initialization, as changing this data *should* trigger a recompile of the script.